File: programmas/lv2/findprime.py
# 소수찾기
from itertools import permutations

def solution(numbers):
    answer = []
    for i in range(1, len(numbers)+1):
        # 순열을 통해 list 생성
        lst = list(permutations(numbers, i))
        # print(lst)
        #lst안에 값들은 튜플로 되어 있음 [('1',), ('7',)] / [('1', '7'), ('7', '1')]
        for j in range(len(lst)):
            # 문자열로 합쳐준 후에 int로 변환
            num = int("".join(lst[j]))
            
            if is_prime(num):
                answer.append(num)
    
    # 중복제거
    answer = list(set(answer))
    
    return len(answer)

# 소수판별
def is_prime(n):
    if n == 0 or n == 1:
        return False
    # 시간복잡도를 조금이라도 올리기 위해 n**0.5 까지만 반복
    for i in range(2, int(n**0.5)+1):
        if n % i == 0:
            return False
    return True


print(solution("011"))

File: programmas/java/lv2/FIndPrime.java
package programmas.java.lv2;

import java.util.HashSet;
import java.util.Set;

public class FindPrime {
    // 1. 순열 생성 -> 소수확인
    // 2. 소수 확인 방법 -> 제곱근까지 곱했을 때 나눠지는 수가 없음
    // 3. 같은 값일 수 있으니 Set에 넣기
    public int solution(String numbers) {
        int result = 0;
        char[] nums = numbers.toCharArray();
        HashSet<Integer> numSet = new HashSet<>();
        boolean[] visited = new boolean[nums.length];

        generatePermutations(nums, new StringBuilder(), visited, numSet);

        for (int num : numSet) {
            if (isPrime(num))
                result++;
        }
        return result;
    }

    private void generatePermutations(char[] nums, StringBuilder curr, boolean[] visited, Set<Integer> numSet) {
        if (curr.length() > 0) {
            numSet.add(Integer.valueOf(curr.toString()));
        }

        for (int i = 0; i < nums.length; i++) {
            if (!visited[i]) {
                visited[i] = true;
                curr.append(nums[i]);
                generatePermutations(nums, curr, visited, numSet);
                curr.deleteCharAt(curr.length() - 1);
                visited[i] = false;
            }
        }
    }

    private boolean isPrime(int num) {
        if (num < 2)
            return false;

        for (int i = 2; i < (int) (Math.sqrt(num)) + 1; i++) {
            if (num % i == 0)
                return false;
        }
        return true;
    }
}
